### ISSUE TITLE
while sourcing nvm command works, npm and node command are not working properly

### ISSUE THREAD

**Issue by shangsunset** [edc/bass#28]
Hi,
So ive put the `nvm` function in the `config.fish` and installed `node` with nvm. however, i always get `fish: Unknown command 'node'` on shell startup, so is `npm`. they will work after i execute a `nvm` command though, such as `nvm ls`.

if `nvm` is already sourced, how come `node` wont work?

ps: tested in bash. node and npm work normally.

my config.fish:

```
# Fisherman config
set fisher_home ~/.local/share/fisherman
set fisher_config ~/.config/fisherman
source $fisher_home/config.fish

# Initial config file contents, which adds /usr/local/bin to the PATH environment variable.
set -g -x PATH /usr/local/bin $PATH

# Homebrew
set -g -x PATH /usr/local/sbin $PATH

# pyenv
status --is-interactive; and . (pyenv init -|psub)

function nvm
    bass source ~/.nvm/nvm.sh ';' nvm $argv
end



# Will eliminate the Fish welcome message that appears by default
set -g -x fish_greeting ''

```

**Comment by bucaran** [comment]
@shangsunset I think the correct function is:

```
function nvm
     bass source ~/.nvm/nvm.sh --no-use ';' nvm $argv
end
```

**Comment by edc** [comment]
@shangsunset let me know if what @bucaran suggested works for you.

**Comment by shangsunset** [comment]
unfortunately, it didnt work.

**Comment by bucaran** [comment]
@shangsunset

Why 

```
set -g -x PATH /usr/local/bin $PATH
```

Isn't /usr/local/bin/ supposed to be in PATH by default?

**Comment by shangsunset** [comment]
Hey @bucaran, I deleted `set -g -x PATH /usr/local/bin $PATH` and this is the output from running `printf "%s\n" $PATH`:

```
/Users/shangsunset/.pyenv/shims
/usr/local/sbin
/Users/shangsunset/.nvm/versions/node/v5.10.1/bin
/usr/local/bin
/usr/bin
/bin
/usr/sbin
/sbin
```

**Comment by bucaran** [comment]
`/Users/shangsunset/.nvm/versions/node/v5.10.1/bin` shows first than `/usr/local/bin`, which makes sense, but what's inside the `nvm` path?

```
ls -la /Users/shangsunset/.nvm/versions/node/v5.10.1/bin
```

**Comment by shangsunset** [comment]
```
total 46224
drwxr-xr-x   4 shangsunset  staff       136 Apr  5 17:47 .
drwxr-xr-x  10 shangsunset  staff       340 Apr 15 09:34 ..
-rwxr-xr-x   1 shangsunset  staff  23661896 Apr  5 17:47 node
lrwxr-xr-x   1 shangsunset  staff        38 Apr  5 17:47 npm -> ../lib/node_modules/npm/bin/npm-cli.js
```

**Comment by bucaran** [comment]
And you are saying that when you run, `nmp` or `node`, you get a "command not found" error?

**Comment by shangsunset** [comment]
yes

**Comment by bucaran** [comment]
And what happens if you do

```
/Users/shangsunset/.nvm/versions/node/v5.10.1/bin/node -v
```

**Comment by bucaran** [comment]
Also, do you have a .bash_profile file in your `$HOME` directory? and if you do, what is its contents?

**Comment by shangsunset** [comment]
output from `/Users/shangsunset/.nvm/versions/node/v5.10.1/bin/node -v`:
```
v5.10.
```

.bash_profile:
```
export NVM_DIR="$HOME/.nvm"
[ -s "$NVM_DIR/nvm.sh" ] && . "$NVM_DIR/nvm.sh" # This loads nvm
```

does .bash_profile still matter if we are using fish?

**Comment by bucaran** [comment]
It does not matter in fish, but nvm is a bash script and I wanted to see if you were defining any aliases.

So, if `/Users/shangsunset/.nvm/versions/node/v5.10.1/bin/node` works, how come `node`, does not work?

It seems to me there is a deeper problem here.

**Comment by bucaran** [comment]
Right, it does not matter in fish, but nvm is a bash script and I wanted to see if you were defining any aliases.

So, if `/Users/shangsunset/.nvm/versions/node/v5.10.1/bin/node` works, how come `node`, does not?

It seems to me there is a deeper problem here.


EDIT: Just to try something:

```
set -gx PATH /Users/shangsunset/.nvm/versions/node/v5.10.1/bin
```

And try running `node` again.

**Comment by shangsunset** [comment]
mmm that works

**Comment by bucaran** [comment]
Check what's inside:

```
/Users/shangsunset/.pyenv/shims
/usr/local/sbin
```

**Comment by shangsunset** [comment]
```
shangsunset@Yeshens-MBP ~> ls -la /Users/shangsunset/.pyenv/shims
total 0
drwxr-xr-x  2 shangsunset  staff   68 Apr 16 00:31 .
drwxr-xr-x  4 shangsunset  staff  136 Apr 11 10:50 ..
shangsunset@Yeshens-MBP ~> ls -la /usr/local/sbin
total 16
drwxr-xr-x   4 shangsunset  admin  136 Apr 12 09:04 .
drwxrwxr-x  20 shangsunset  admin  680 Apr 12 10:07 ..
lrwxr-xr-x   1 shangsunset  admin   35 Apr 12 09:04 php-fpm -> ../Cellar/php56/5.6.20/sbin/php-fpm
lrwxr-xr-x   1 shangsunset  admin   37 Apr 12 09:04 php56-fpm -> ../Cellar/php56/5.6.20/sbin/php56-fpm
shangsunset@Yeshens-MBP ~>
```

**Comment by bucaran** [comment]
Well, that looks alright, I guess.

Comment out all of your config.fish, and just leave this

```
set fisher_home ~/.local/share/fisherman
set fisher_config ~/.config/fisherman
source $fisher_home/config.fish

function nvm
   bass source ~/.nvm/nvm.sh --no-use ';' nvm $argv
end
```

Then reload the shell and try again.

**Comment by shangsunset** [comment]
still getting `fish: Unknown command 'node'` :(

**Comment by bucaran** [comment]
And if you modify your PATH like this?

```
/Users/shangsunset/.nvm/versions/node/v5.10.1/bin
/usr/local/bin
/usr/bin
/bin
/usr/sbin
/sbin
```

**Comment by shangsunset** [comment]
yea node works with `/Users/shangsunset/.nvm/versions/node/v5.10.1/bin` in PATH

**Comment by bucaran** [comment]
So, the problem is solved. What is the issue?

**Comment by shangsunset** [comment]
so i need to put `set -gx PATH /Users/shangsunset/.nvm/versions/node/v5.10.1/bin $PATH` in my config.fish?

**Comment by bucaran** [comment]
@shangsunset Actually, not, that's exactly `nvm`'s job. So, what you need to do is run `nvm` or `nvm use foo` when you restart your shell / cd into a path with a `.nvmrc`, etc.

**Comment by bucaran** [comment]
@shangsunset 

Nope, that's exactly `nvm`'s job. So, what you need to do is run `nvm` or `nvm use foo` when you restart your shell / cd into a path with a `.nvmrc`, etc.

Note that you need to do this **everytime**, because `nvm` modifcations in your PATH will not be persisted across sessions.

What's the aftermath of all this? Were you using `nvm` incorrectly?

**Comment by bucaran** [comment]
@shangsunset 

Nope, that's exactly `nvm`'s job. You need to do is run `nvm` or `nvm use foo` when you restart your shell / cd into a path with a `.nvmrc`, etc., **everytime**.

Because `nvm` mods in your PATH will not be persisted across sessions AFIK.

I think you were just using `nvm` incorrectly, so not your fault, I was confused about this too.

**Comment by bucaran** [comment]
@shangsunset 

Nope, that's exactly `nvm`'s job. You need to do is run `nvm` or `nvm use foo` when you restart your shell / cd into a path with a `.nvmrc`, etc., **everytime**.

Because `nvm` mods in your PATH will not be persisted across sessions AFIK.

I think you were just using `nvm` incorrectly, so not your fault, I was confused about this too.

**Comment by bucaran** [comment]
@shangsunset 

Nope, that's exactly `nvm`'s job. You need to do is run `nvm` or `nvm use foo` when you restart your shell / cd into a path with a `.nvmrc`, etc., **everytime**.

Because `nvm` mods in your PATH will not be persisted across sessions AFIK.

I think you were just using `nvm` incorrectly, so not your fault, I was confused about this too.

**Comment by bucaran** [comment]
@shangsunset 

Nope, that's exactly `nvm`'s job. You need to do is run `nvm` or `nvm use foo` when you restart your shell / cd into a path with a `.nvmrc`, etc., **everytime**.

Because `nvm` mods in your PATH will not be persisted across sessions AFIK.

I think you were just using `nvm` incorrectly, so not your fault, I was confused about this too.

**Comment by bucaran** [comment]
@shangsunset 

Nope, that's exactly `nvm`'s job. You need to do is run `nvm` or `nvm use foo` when you restart your shell / cd into a path with a `.nvmrc`, etc., **everytime**.

Because `nvm` mods in your PATH will not be persisted across sessions AFIK.

I think you were just using `nvm` incorrectly, so not your fault, I was confused about this too.

**Comment by bucaran** [comment]
@shangsunset 

Nope, that's exactly `nvm`'s job. You need to do is run `nvm` or `nvm use foo` when you restart your shell / cd into a path with a `.nvmrc`, etc., **everytime**.

Because `nvm` mods in your PATH will not be persisted across sessions AFIK.

I think you were just using `nvm` incorrectly, so not your fault, I was confused about this too.

**Comment by shangsunset** [comment]
I need to run `nvm` every time when i use `node` or `npm` from now on? That doesnt sound very ideal...
I wonder why I didnt have to run `nvm` every time when i was using `zsh` or `omf/plugin-nvm`. hmmmm...

**Comment by bucaran** [comment]
@shangsunset If you use `nvm` like a function, you need to run it **everytime**, if you don't want to run it everytime, you can **source** `nvm` in your config.fish or create a snippet that does it for you.

I already answered how you can accmplish all that here: https://github.com/creationix/nvm/issues/303#issuecomment-210738874

**Comment by bucaran** [comment]
@shangsunset 

Just `rm ~/.config/functions/config.fish`, and create a new one with this:

```
set fisher_home ~/.local/share/fisherman
set fisher_config ~/.config/fisherman
source $fisher_home/config.fish

set -g -x PATH /usr/local/sbin $PATH
status --is-interactive; and . (pyenv init -|psub)

##########################################################

bass source ~/.nvm/nvm.sh ';' nvm $argv

##########################################################
```

**Comment by bucaran** [comment]
@shangsunset 

Just `rm ~/.config/functions/config.fish`, and create a new `config.fish` with this:

```fish
set fisher_home ~/.local/share/fisherman
set fisher_config ~/.config/fisherman
source $fisher_home/config.fish

set -g -x PATH /usr/local/sbin $PATH
status --is-interactive; and . (pyenv init -|psub)


### define a nvm wrapper function so you can call it whenever you want
function nvm
     bass source ~/.nvm/nvm.sh --no-use ';' nvm $argv
end

### call nvm HERE so that there is no need to call it every time
nvm 
```

**Comment by bucaran** [comment]
@shangsunset 

Just `rm ~/.config/functions/config.fish`, and create a new `config.fish` with this:

```fish
set fisher_home ~/.local/share/fisherman
set fisher_config ~/.config/fisherman
source $fisher_home/config.fish

set -g -x PATH /usr/local/sbin $PATH
status --is-interactive; and . (pyenv init -|psub)


### define a nvm wrapper function so you can call it whenever you want
function nvm
     bass source ~/.nvm/nvm.sh --no-use ';' nvm $argv
end

### call nvm HERE so that there is no need to call it every time
nvm 
```

**Comment by bucaran** [comment]
@shangsunset 

Just `rm ~/.config/functions/config.fish`, and create a new `config.fish` with this:

```fish
set fisher_home ~/.local/share/fisherman
set fisher_config ~/.config/fisherman
source $fisher_home/config.fish

set -g -x PATH /usr/local/sbin $PATH
status --is-interactive; and . (pyenv init -|psub)


### define a nvm wrapper function so you can call it whenever you want
function nvm
     bass source ~/.nvm/nvm.sh --no-use ';' nvm $argv
end

### call nvm HERE so that there is no need to call it every time
nvm 
```

**Comment by bucaran** [comment]
@shangsunset 

Just `rm ~/.config/functions/config.fish`, and create a new `config.fish` with this:

```fish
set fisher_home ~/.local/share/fisherman
set fisher_config ~/.config/fisherman
source $fisher_home/config.fish

set -g -x PATH /usr/local/sbin $PATH
status --is-interactive; and . (pyenv init -|psub)


### define a nvm wrapper function so you can call it whenever you want
function nvm
     bass source ~/.nvm/nvm.sh --no-use ';' nvm $argv
end

### call nvm HERE so that there is no need to call it every time
nvm 
```

**Comment by bucaran** [comment]
@shangsunset 

Just `rm ~/.config/functions/config.fish`, and create a new `config.fish` with this:

```fish
set fisher_home ~/.local/share/fisherman
set fisher_config ~/.config/fisherman
source $fisher_home/config.fish

set -g -x PATH /usr/local/sbin $PATH
status --is-interactive; and . (pyenv init -|psub)


### define a nvm wrapper function so you can call it whenever you want
function nvm
     bass source ~/.nvm/nvm.sh --no-use ';' nvm $argv
end

### call nvm HERE so that there is no need to call it every time
nvm 
```

**Comment by bucaran** [comment]
@shangsunset 

Just `rm ~/.config/functions/config.fish`, and create a new `config.fish` with this:

```fish
set fisher_home ~/.local/share/fisherman
set fisher_config ~/.config/fisherman
source $fisher_home/config.fish

set -g -x PATH /usr/local/sbin $PATH
status --is-interactive; and . (pyenv init -|psub)


### define a nvm wrapper function so you can call it whenever you want
function nvm
     bass source ~/.nvm/nvm.sh --no-use ';' nvm $argv
end

### call nvm HERE so that there is no need to call it every time
nvm 
```

**Comment by bucaran** [comment]
@shangsunset 

Just `rm ~/.config/functions/config.fish`, and create a new `config.fish` with this:

```fish
set fisher_home ~/.local/share/fisherman
set fisher_config ~/.config/fisherman
source $fisher_home/config.fish

set -g -x PATH /usr/local/sbin $PATH
status --is-interactive; and . (pyenv init -|psub)


### define a nvm wrapper function so you can call it whenever you want
function nvm
     bass source ~/.nvm/nvm.sh --no-use ';' nvm $argv
end

### call nvm HERE so that there is no need to call it every time
nvm 
```

**Comment by bucaran** [comment]
@shangsunset 

Just `rm ~/.config/functions/config.fish`, and create a new `config.fish` with this:

```fish
set fisher_home ~/.local/share/fisherman
set fisher_config ~/.config/fisherman
source $fisher_home/config.fish

### add your other configuration here

function nvm
     bass source ~/.nvm/nvm.sh --no-use ';' nvm $argv
end

nvm  ### call nvm HERE so you don't have to call it every time

```

**Comment by shangsunset** [comment]
@bucaran 

Really appreciate your help and patience. However, even though i deleted my existing config.fish and created a new one with what you provided, `node` and `npm` are still not available upon shell start. unless i run `nvm use node`.

**Comment by bucaran** [comment]
@shangsunset Why are you using nvm? I think you could make your life easier if you switched to tj/n. The use case there looks more suited to you.

https://github.com/tj/n

**Comment by shangsunset** [comment]
ok thanks.

**Comment by bucaran** [comment]
@shangsunset Built this just for you

https://github.com/fisherman/nvm

**Comment by ljharb** [comment]
Please do not steal other people's project names.

**Comment by bucaran** [comment]
@ljharb Dude, you are taking this the wrong way. The purpose of fisherman/nvm is to serve as a bridge between creationix/nvm and the missing nvm solution in fish. Why couldn't you think of that before thinking I was stealing your project's "name"?

**Comment by ljharb** [comment]
I did - I asked you politely to change it. You said "No". call it `nvm-fish` if you like - but not "nvm". That's confusing for users, and makes googling harder, even if it *wasn't* rude and potentially an implicit trademark violation.

**Comment by bucaran** [comment]
People that use fisherman know stuff inside the org are all plugins for fish, kind of like ohmyfish and similar projects. I wouldn't say it's confusing.

**Comment by ljharb** [comment]
You were politely asked by a project whose name you copied to change it. Whether you agree that it's confusing is irrelevant - it's not cool, under any circumstances.

**Comment by bucaran** [comment]
First, you were **not** polite, you were not rude either, but you were definitely very hostile. You even called me an "asshole". Why should I accept that? People on the internet are very confrontational when facing a message box. I wish you would have called me an asshole in my face...

Second, I worked hard on this thing, not that it matters, but I put it there because a lot of people have asked for an nvm in fish for like years and there just wasn't one. Sure, there are wrappers, but it's just not the same. I am calling the project nvm for a simple reason, people migrating to fish that have used nvm before will likely search for "nvm in fish" when they realize creationix/nvm does not work in fish.

**Comment by ljharb** [comment]
so call it "nvm in fish". I insulted you after you posted a *very* rude one-word answer "No", and after you called me "bro" (which is also rude; you don't know me) and told me to "chill".

I'm glad you worked hard on it, and I hope your project is successful - but it should be successful under its own name, not by stealing the name of something else. If you don't change your name, and your project attains any notoriety, it will forever live under the shadow of that - I won't be the only one you hear it from. Why not just change it?

**Comment by bucaran** [comment]
Well, that's a touching way to put it. There you go: https://github.com/fisherman/nbm ;)

And, I said "dude". From there to asshole it's quite a big jump.

**Comment by ljharb** [comment]
Thanks, much obliged.

**Comment by bucaran** [comment]
@ljharb Ping ↑

**Comment by ljharb** [comment]
@bucaran  thank you, that sounds great. Please feel free to send a PR.

**Comment by FabioAntunes** [comment]
@bucaran fish-nvm was way better than fin, it's very funny naming stuff after fishes and stuff related to them, but that makes googling hard as hell.

If I want a nvm for fish I will search for that on google, not fin. If I want to run bash script on my fish shell I will search for fish bash wrapper or fish bash runner, not bass.

@shangsunset You can solve your problem sourcing nvm every time you open a new shell, this will add a minor delay, but at least it works, just put this line on your config file at ~/.config/fish/config.fish 

`bass source ~/.nvm/nvm.sh >/dev/null ^&1`

**Comment by bucaran** [comment]
@FabioAntunes Which is why I named it nvm in the first place, but then ljharb was right, I can't just name my project after nvm because we are not affiliated so I changed it to fin. 

fin is not a drop-in replacement for nvm nor it requires `nvm` as a dependency, which actually threw off a few people at first, ping @amir20.

> If I want a nvm for fish I will search for that on google, not fin. If I want to run bash script on my fish shell I will search for fish bash wrapper or fish bash runner, not bass.

Yeah, I couldn't agree more.
